docs/nsm_api.md

```diff
@@ -0,0 +1,445 @@
+
+Network Service Mesh is orthogonal to CNI and normal Kubernetes Networking.  It must not interfere with or impede normal Kubernetes Networking functionality.
+
+## Network Service Mesh APIs
+
+### Network Service Mesh APIs in the Abstract
+
+In the abstract, the APIs that an Network Service Manager uses to communicate with the Network Service Clients, Network Service Endpoints, and Network Service Dataplanes in its Network Service Manager Domain can be anything that works well within that Network Service Manager Domain.  It is expected that over time, standards for particular kinds of Network Service Manager Domains will emerge.  In subsequent sections, this document will define such a standard for Network Service Managers within the domain of a Kubernetes Node.  
+
+Additionally, the mechanisms for Network Service Registries in a Network Service Registry domain can also be whatever works for that Network Service Registry Domain and the Network Service Managers within it.  In subsequent sections, this document will define such a standard for Network Service Registries for Kubernetes Clusters.
+
+This means that in the abstract Network Service Mesh has:
+- A concrete Network Service Manager to Network Service Manager API
+- A logical Network Service Registry API - that information that must be stored in the Network Service Registry so Network Service Managers function.
+
+#### Network Service Manager to Network Service Manager (NSM2NSM) API
+
+![NSM2NSM API](./images/NSM2NSM.png)
+
+The Network Service Manager API is defined in protobuf for communication over grpc.
+
+In this message Client&#39;s local NSM proxying Network Service request to NSE&#39;s local NSM.  2 NSMs need to agree upon a tunneling technology supported by both, hence a list of tunnel types is a part of the request message.
+
+```proto
+message RemoteConnectionRequest {
+   required string request_id = 1;
+   optional string network_registry_name = 2; // Do we want to have a way to identify the registry domain here?
+   required string network_service_name = 3;
+   string nse_provider_name = 4;
+   repeated RemoteMechanismRequest tunnel_type = 5;
+}
+
+message RemoteConnectionReply {
+    required string request_id = 1;
+    required bool accepted = 2;
+    optional admission_error = 3;
+    required RemoteMechanismReply = 4;
+}
+
+/* 
+ *  Initial attempt to define remote mechanism types.  Very preliminary
+ */
+enum RemoteMechanismType {
+    NONE = 1; // For use when connection is not accepted
+    VXLAN = 2;
+    VXLAN-GPE = 3;
+    GRE = 4;
+    MPLSoEthernet = 5;
+    MPLSoGRE = 6;
+    MPLSoUDP = 7;
+}
+
+/*
+ * Vxlan used as example, others will have to be filled out
+ */
+message Vxlan_Parameters {
+    required bytes src_ip = 1;
+    required bytes dst_ip = 2;
+    optional bytes src_port = 3;
+    optional bytes dst_port = 4;
+    required unint32 vni = 5;
+}
+
+message RemoteMechanismParameters {
+    oneof {
+        None_Parameters = 1;
```

Comment:
It should be defined as a separate ENUM, for reuse in other messages.

Reply:
I hadn't finished this quite yet... this is a oneof, not an ENUM... what its basically going to say (when I finish) is that this can be one of the various parameter types for the various remote mechanism types.